pkg/kubelet/pleg/pleg.go

```diff
@@ -63,9 +63,14 @@ type PodLifecycleEvent struct {
 // PodLifecycleEventGenerator contains functions for generating pod life cycle events.
 type PodLifecycleEventGenerator interface {
 	Start()
-	Stop()
-	Update(relistDuration *RelistDuration)
 	Watch() chan *PodLifecycleEvent
 	Healthy() (bool, error)
+	PodLifecycleEventGeneratorHandler
```

Comment:
@harche Thank you for your PR!

As far as I understood [@derekwaynecarr's comment](https://github.com/kubernetes/kubernetes/pull/111384#discussion_r1015933735) `PodLifecycleEventGenerator` should not include `PodLifecycleEventGeneratorHandler` methods as `this is all the rest of the kubelet needs to know about`. I guess you should do it other way around - `PodLifecycleEventGeneratorHandler` should embed  `PodLifecycleEventGenerator`.

Reply:
@harche please, address this comment to proceed further with the PR, thanks.

Reply:
I will think through this and reply here. Thanks.

Reply:
`GenericPLEG` had been an implementation of the [PodLifecycleEventGenerator](https://github.com/kubernetes/kubernetes/blob/ed8cad1e80d096257921908a52ac69cf1f41a098/pkg/kubelet/pleg/generic.go#L119) interface and so is [recently added EventedPLEG.](https://github.com/kubernetes/kubernetes/blob/ed8cad1e80d096257921908a52ac69cf1f41a098/pkg/kubelet/pleg/evented.go#L90). I am trying to accomodate @derekwaynecarr's request to seperate the newly added methods from existing methods without bringing in any major changes to how is it designed (especially `GenericPLEG`) as it stands today. 

What do you think of following approach where we reduce the scope of `podLifecycleEventGeneratorHandler` by making it private?

```golang

type PodLifecycleEventGenerator interface {
	Start()
	Watch() chan *PodLifecycleEvent
	Healthy() (bool, error)
	podLifecycleEventGeneratorHandler
}

type podLifecycleEventGeneratorHandler interface {
	Stop()
	Update(relistDuration *RelistDuration)
	Relist()
}
```

Reply:
Although the moment you put `podLifecycleEventGeneratorHandler` inside `PodLifecycleEventGenerator` it is no longer private.

Reply:
I'd suggest to do it other way around.

Quoting @derekwaynecarr:

> PodLifecycleEventGenerator stays with just Start, Watch, and Healthy as this is all the rest of the kubelet needs to know about.
```
type PodLifecycleEventGenerator interface {
	Start()
	Watch() chan *PodLifecycleEvent
	Healthy() (bool, error)
}
```

> We add a PodLifecycleEventGeneratorHandler and it adds Stop , Update, and Relist so these compositions can be kept internal.
```
type PodLifecycleEventGeneratorHandler interface {
	PodLifecycleEventGenerator
        Stop()
	Update(relistDuration *RelistDuration)
	Relist()
}
```

Reply:
If we go this way then we will have to return `PodLifecycleEventGeneratorHandler` instead of `PodLifecycleEventGenerator` when initializing PLEG, https://github.com/kubernetes/kubernetes/blob/ed8cad1e80d096257921908a52ac69cf1f41a098/pkg/kubelet/pleg/generic.go#L119

Other way of saying this would be, from now on any PLEG will be an implementation of `PodLifecycleEventGeneratorHandler` instead of `PodLifecycleEventGenerator`. This way will still expose `Stop()`, `Update()` and `Relist()` for anyone who has gotten hold of the PLEG implementation.

Reply:
As far as I can see In the current code only Start, Watch and Healthy methods are used outside of pkg/kubelet/pleg module. Could it be the reason why @derekwaynecarr was asking to move the rest of the methods to another interface?

Reply:
@harche Any news? Are you going to continue working on this?

Reply:
I am not sure if we have a consensus based on the discussion above.

Reply:
Let's discuss it again if it's still relevant. I quoted @derekwaynecarr above. He explicitly stated which interface should have which methods. Am I missing something obvious here?

Reply:
Comments, https://github.com/kubernetes/kubernetes/pull/113825#discussion_r1081448798, https://github.com/kubernetes/kubernetes/pull/113825#discussion_r1081456783 and https://github.com/kubernetes/kubernetes/pull/113825#discussion_r1082665921 describe the issue I am having implmenting this without making major changes to existing `Generic PLEG`. 

I'm open to discussing alternative approaches or further clarifying my concerns if needed.

Reply:
Nope, it's not right. The latest changes don't implement the interface correctly. Let me see we can fix it.

Reply:
Well the interdepencies between those 6 methods is such that, if you try to truly separate them in two different interfaces it starts to mess around with the simple design of the existing Generic PLEG. I am not sure if this refactor is worth that. 

But again, may be I am overlooking a simpler approach, so I am very much open to discussing the alternative approach.

Reply:
@harche FWIW, here is what I came up with. Not sure if this would work, but it compiles and passes all Kubelet unit tests. WDYT?
```
diff --git a/pkg/kubelet/pleg/evented.go b/pkg/kubelet/pleg/evented.go
index 1dd176489a5..704e9e5390a 100644
--- a/pkg/kubelet/pleg/evented.go
+++ b/pkg/kubelet/pleg/evented.go
@@ -71,7 +71,7 @@ type EventedPLEG struct {
        // For testability.
        clock clock.Clock
        // GenericPLEG is used to force relist when required.
-       genericPleg PodLifecycleEventGenerator
+       genericPleg PodLifecycleEventGeneratorHandler
        // The maximum number of retries when getting container events from the runtime.
        eventedPlegMaxStreamRetries int
        // Indicates relisting related parameters
@@ -93,7 +93,7 @@ func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.Ru
                runtimeService:              runtimeService,
                eventChannel:                eventChannel,
                cache:                       cache,
-               genericPleg:                 genericPleg,
+               genericPleg:                 genericPleg.(PodLifecycleEventGeneratorHandler),
                eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
                relistDuration:              relistDuration,
                clock:                       clock,
diff --git a/pkg/kubelet/pleg/pleg.go b/pkg/kubelet/pleg/pleg.go
index 2654f32d6fc..f972cc10f86 100644
--- a/pkg/kubelet/pleg/pleg.go
+++ b/pkg/kubelet/pleg/pleg.go
@@ -64,10 +64,14 @@ type PodLifecycleEvent struct {
 // PodLifecycleEventGenerator contains functions for generating pod life cycle events.
 type PodLifecycleEventGenerator interface {
        Start()
-       Stop()
-       Update(relistDuration *RelistDuration)
        Watch() chan *PodLifecycleEvent
        Healthy() (bool, error)
-       Relist()
        UpdateCache(*kubecontainer.Pod, types.UID) (error, bool)
 }
+
+type PodLifecycleEventGeneratorHandler interface {
+       PodLifecycleEventGenerator
+       Stop()
+       Update(relistDuration *RelistDuration)
+       Relist()
+}
```